### PR TITLE
Add Slot type to abstract the raw pointers being used for slots.

### PR DESCRIPTION
**Pull request by zdevito**
If modules become first-class IValues, then the slots will no longer be raw pointers but (IValue, index) pairs. This commit inserts the Slot abstraction so that this change can be made in later patches.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#18226 Add Slot type to abstract the raw pointers being used for slots.**

Differential Revision: [D14542022](https://our.internmc.facebook.com/intern/diff/D14542022)

### CODE REVIEW

File: torch/csrc/jit/script/slot.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  }
+  at::IValue* operator->() const {
+      return slot_;
+  }
+  bool operator==(const Slot& rhs) const {
+    return slot_ == rhs.slot_;
+  }
+private:
+  at::IValue* slot_;
+  friend struct std::hash<Slot>;
```
Comment (Krovatkin) [inline torch/csrc/jit/script/slot.h]: I don't know if `return std::hash<at::IValue*>{}(&(*s.slot_));` is any better than declaring a friend.

File: torch/csrc/jit/script/module.h [hunk 2]
```diff
@@ -361,15 +363,22 @@ struct NamedModule {
...
       : name_(name),
-        type(type),
-        ivalue(torch::make_unique<IValue>(std::move(ivalue))) {}
+        type_(type),
+        ivalue_(torch::make_unique<IValue>(std::move(ivalue))) {}
 
-  IValue* slot() const {
-    return ivalue.get();
+  Slot slot() const {
+    return Slot(ivalue_.get());
```
Comment (Krovatkin) [inline torch/csrc/jit/script/module.h]: Oh, I see now, the lifetimes of `IValue` and `Slot` are  the same unless someone creates the `Slot` outside of a `NamedValue`. Would it make sense to make `Slot` c-tor private, so only `NamedValue`s can create `Slot`s?

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: torch/csrc/jit/script/slot.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    return *slot_;
+  }
+  at::IValue* operator->() const {
+      return slot_;
+  }
+  bool operator==(const Slot& rhs) const {
+    return slot_ == rhs.slot_;
+  }
+private:
+  at::IValue* slot_;
```
[the remark this answers is not in the document]
Comment (zdevito) [reply torch/csrc/jit/script/slot.h]: The IValue* lives as long as the module is alive.

File: torch/csrc/jit/script/slot.h (hunk 1 above)
Comment (zdevito) [reply torch/csrc/jit/script/slot.h]: <in person>: This particular implementation will only live until a later diff in this stack merges Slot and NamedIValue, so I will keep it as is for now.

**Pull request closed without merging**